korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt

```diff
@@ -41,13 +47,82 @@ interface BmpCoordsWithT<T : ISizeInt> : BmpCoords, Closeable, Resourceable<BmpC
     val width: Int get() = (Point.distance(tl_x * baseWidth, tl_y * baseHeight, tr_x * baseWidth, tr_y * baseHeight)).toInt()
     val height: Int get() = (Point.distance(tl_x * baseWidth, tl_y * baseHeight, bl_x * baseWidth, bl_y * baseHeight)).toInt()
-    val frameOffsetX: Int get() = 0
-    val frameOffsetY: Int get() = 0
-    val frameWidth: Int get() = width
-    val frameHeight: Int get() = height
+    val virtFrame: RectangleInt? get() = null
+    val frameOffsetX: Int get() = virtFrame?.x ?: 0
+    val frameOffsetY: Int get() = virtFrame?.y ?: 0
+    val frameWidth: Int get() = virtFrame?.width ?: width
+    val frameHeight: Int get() = virtFrame?.height ?: height
     val area: Int get() = width * height
+    val isRotatedInBaseDeg90 get() = (width > 1 && tl_x == tr_x) || (height > 1 && tl_y == bl_y)
     override fun close() = Unit
     val sizeString: String get() = "${width}x${height}"
+
+    fun subCoords(bounds: RectangleInt, imageOrientation: ImageOrientation): BmpCoordsWithT<T> {
+        // Assure bounds are inside this slide
+        if (bounds.x < 0) {
+            bounds.width += bounds.x
+            bounds.x = 0
+        }
+        if (bounds.y < 0) {
+            bounds.height += bounds.y
+            bounds.y = 0
+        }
+        if (imageOrientation.isRotatedDeg90CwOrCcw) {
+            bounds.width = bounds.width.clamp(0, height - bounds.y)
+            bounds.height = bounds.height.clamp(0, width - bounds.x)
+        } else {
+            bounds.width = bounds.width.clamp(0, width - bounds.x)
+            bounds.height = bounds.height.clamp(0, height - bounds.y)
+        }
+
+        val tlX: Float
+        val tlY: Float
+        val trX: Float
+        val trY: Float
+        val brX: Float
+        val brY: Float
+        val blX: Float
+        val blY: Float
+
+        // Calculate bmpCoords based on parentCoords
+        val dx = br_x - tl_x
+        val dy = br_y - tl_y
+        val x = bounds.x.toFloat()
+        val y = bounds.y.toFloat()
+        val w = bounds.width.toFloat()
+        val h = bounds.height.toFloat()
+        val bw = width.toFloat()
+        val bh = height.toFloat()
+
+        if (isRotatedInBaseDeg90) {
+            // Base is rotated
+            val xDim = if (imageOrientation.isRotatedDeg90CwOrCcw) w else h
+            val yDim = if (imageOrientation.isRotatedDeg90CwOrCcw) h else w
+            tlX = tl_x + y / bh * dx
+            tlY = tl_y + x / bw * dy
+            brX = tlX + xDim / bh * dx
+            brY = tlY + yDim / bw * dy
+            trX = tlX
+            trY = brY
+            blX = brX
+            blY = tlY
+        } else {
+            // Base is not rotated
+            val xDim = if (imageOrientation.isRotatedDeg90CwOrCcw) h else w
+            val yDim = if (imageOrientation.isRotatedDeg90CwOrCcw) w else h
+            tlX = tl_x + x / bw * dx
+            tlY = tl_y + y / bh * dy
+            brX = tlX + xDim / bw * dx
+            brY = tlY + yDim / bh * dy
```

Comment:
These lines look duplicated (only nverting bw/bh)

Reply:
Not really, all lines are different. `x`, `y`, `h` and `w` are also inverted and for example `trX` is assigned from `tlX` or `brX`

Reply:
I was thinking on something like this (untested, so not sure if I flipped something): 

```kotlin
val flipDim = isRotatedInBaseDeg90 xor imageOrientation.isRotatedDeg90CwOrCcw
val xDim = if (flipDim) h else w
val yDim = if (flipDim) w else h
val _bw = if (isRotatedInBaseDeg90) bh else bw
val _bh = if (isRotatedInBaseDeg90) bw else bh
val _x = if (isRotatedInBaseDeg90) y else x
val _y = if (isRotatedInBaseDeg90) x else y
tlX = tl_x + _x / _bw * dx
tlY = tl_y + _y / _bh * dy
brX = tlX + xDim / _bh * dx
brY = tlY + yDim / _bh * dy
```

But that way the arithmetic formulas are not repeated